// rxjs-observe/esm2015/observe.js
import { BehaviorSubject, noop, Subject } from "rxjs";
export function observe(instance, callbacks) {
    const defaultedCallbacks = callbacks || {};
    const subjects = new Map();
    const proxy = new Proxy(instance, {
        get(target, name) {
            const callbacksValue = defaultedCallbacks[name];
            const targetValue = target[name];
            let value = callbacksValue && !targetValue ? callbacksValue : targetValue;
            if (typeof value === "function") {
                const functionValue = value;
                value = function (...args) {
                    const result = functionValue.apply(this, args);
                    const subject = subjects.get(name);
                    if (subject) {
                        subject.next(args);
                    }
                    return result;
                };
            }
            return value;
        },
        getOwnPropertyDescriptor(target, name) {
            return (Object.getOwnPropertyDescriptor(target, name) ||
                Object.getOwnPropertyDescriptor(defaultedCallbacks, name));
        },
        has(target, name) {
            return name in target || name in defaultedCallbacks;
        },
        ownKeys(target) {
            return [
                ...Reflect.ownKeys(target),
                ...Reflect.ownKeys(defaultedCallbacks)
            ];
        },
        set(target, name, value) {
            target[name] = value;
            const subject = subjects.get(name);
            if (subject) {
                subject.next(value);
            }
            return true;
        }
    });
    return {
        observables: new Proxy({}, {
            get(target, name) {
                let subject = subjects.get(name);
                if (!subject) {
                    subject =
                        typeof instance[name] === "function" ||
                            typeof defaultedCallbacks[name] === "function"
                            ? new Subject()
                            : new BehaviorSubject(instance[name]);
                    subjects.set(name, subject);
                }
                return subject.asObservable();
            }
        }),
        proxy
    };
}
export function callback() {
    return noop;
}
